src/rate_limiter.cc

```diff
@@ -775,52 +757,89 @@ RateLimiter::ResourceManager::AddModelInstance(
           resource.count();
     }
   }
+  // Increase max resource if needed.
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  UpdateMaxResource(pr.first->second);
+  return ParseAndValidateResources();
 }
 
 Status
 RateLimiter::ResourceManager::RemoveModelInstance(
     const ModelInstanceContext* instance)
 {
-  std::lock_guard<std::mutex> lk(model_resources_mtx_);
+  // Find instance from model resources.
+  std::lock_guard<std::mutex> lk1(model_resources_mtx_);
   const auto& itr = model_resources_.find(instance);
   if (itr == model_resources_.end()) {
     return Status(
         Status::Code::INTERNAL, "Can not find the instance to remove");
   }
+  // Check if max resources need to be updated.
+  bool update_needed = false;
+  std::lock_guard<std::mutex> lk2(max_resources_mtx_);
+  for (const auto& resource_device_map : itr->second) {
+    auto ditr = max_resources_.find(resource_device_map.first);
+    if (ditr != max_resources_.end()) {
+      for (const auto resource : resource_device_map.second) {
+        auto ritr = ditr->second.find(resource.first);
+        if (ritr != ditr->second.end() && ritr->second >= resource.second) {
+          update_needed = true;
+          break;
+        }
+      }
+    }
+    if (update_needed) {
```

Comment:
At a glance from the loops above, it looks like we need to quite a bit of looping. Does it simplify things if the `max_resources_` object also held some list of `ModelInstance*` that contributed to it? 

So upon removal you can just check if the instance is contained in that max resource list and immediately know it needs to be recomputed? Not sure if that overcomplicates things, or the rough size/impact of the loops above. Let me know what you think.

Reply:
I like the idea, but there is an added complexity in managing the contents in "some list of `ModelInstance*`". For example, consider adding instances 1 to 3 with cost `[1, 2, 3]` respectively. When instance 2 is being added, it will see a cost of `1` is being replaced, but not knowing which instance contributed to the cost of `1`. Of course, in this simple example, the list of `ModelInstance*` only contains instance 1, which is "known". In the real world, the max resource is a map from multiple `device_id` to multiple `resource_type_name` and its 'resource limit', which is reasonable to have multiple instances in the list of `ModelInstance*` without knowing exactly which 'resource limit' it is specifying. To address this issue, it is possible to put the pointer next to the 'resource limit', but the 'resource limit' is already sufficient at making the decision without needing the pointer.

Please let me know if I am missing something here.